src/main/java/net/glowstone/entity/GlowLivingEntity.java

```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
     // Health
 
     @Override
-    public <T extends Projectile> T launchProjectile(Class<? extends T> projectile,
-        Vector velocity) {
-        T entity = world.spawn(getEyeLocation(), projectile);
-        entity.setVelocity(velocity);
-        return entity;
+    public <T extends Projectile> T launchProjectile(Class<? extends T> clazz, Vector vector) {
+        float offset = 0.0F;
+        float velocity = 1.5F;
+        T projectile = launchProjectile(clazz, vector, offset, velocity);
+        return projectile;
+    }
+
+    /**
+     * Launches a projectile from this entity.
+     *
+     * @param clazz the projectile class
+     * @param vector TODO: document this parameter
+     * @param offset TODO: document this parameter
+     * @param velocity the velocity for the first flight tick
+     * @param <T> the projectile class
+     * @return the launched projectile
+     */
+    public <T extends Projectile> T launchProjectile(Class<? extends T> clazz, Vector vector,
+            float offset, float velocity) {
+        if (vector == null) {
+            vector = getVelocity();
+        }
+
+        T projectile = throwProjectile(clazz, getEyeLocation().clone(), vector, offset, velocity);
+        projectile.setShooter(this);
+        return projectile;
+    }
+
+    protected <T extends Projectile> T throwProjectile(Class<? extends T> type, Location location,
+            Vector originalVector, float offset, float velocity) {
+        double k = Math.toRadians(-1);
+        double x = cos(k * location.getPitch()) * sin(k * location.getYaw());
```

Comment:
I don't understand your vector math here or in the next method. Could you explain what's being done here?

Reply:
I believe I wrote this math originally in the projectiles branch. Here's my explanation (I learned trigonometry and vector math in another language, sorry if I'm not using the terms properly).

**Pitch/Yaw**
Pitch is defined as the angle of rotation on the horizontal axis, ranging from 90 degrees to -90 degrees.
Yaw is defined as the angle of rotation on the vertical axis ranging from 180 to -180.

**X component**
The first part consists of using the pitch (horizontal-axis rotation angle) to find out at which horizontal angle (downwards/upwards) the player is shooting the projectile. This angle helps as a “multiplier” for the X component, because shooting directly upwards or downwards should nullify the component. Because pitch is ranged negatively (negative is below the origin), we have to negate the conversion:
![image](https://user-images.githubusercontent.com/6775216/35294011-7bcaf83e-0043-11e8-9cbe-71768cf891b3.png)
We must then take into consideration the vertical-axis rotation angle (yaw). A yaw of -90deg represents a full X component (+X), while 90deg represents the opposite (-X). 180 and 0 degrees represent a null X component. This conversion can be achieved by using:
![image](https://user-images.githubusercontent.com/6775216/35294068-a761a628-0043-11e8-9a10-dff0a8b05f5f.png)
Because the pitch acts as a multiplier to the vertical-axis rotation:
![image](https://user-images.githubusercontent.com/6775216/35294084-b6209084-0043-11e8-8218-8d66a5e019cb.png)

**Z component**
This conversion is similar to the X component: it uses the pitch as a multiplier too. What changes is the use of the yaw range, which can be calculated using sin instead of cosine.
![image](https://user-images.githubusercontent.com/6775216/35294149-e250339e-0043-11e8-9896-e3864bd60663.png)

**Y component**
This is similar to the operation used to find the multiplier to the X/Z components, but here we need the upwards and downwards directions to represent a full Y component (using sin) instead of nullifying it (using cosine).
![image](https://user-images.githubusercontent.com/6775216/35294228-01fec70a-0044-11e8-98d0-eb1b4ad22ce3.png)

Reply:
Added a comment to explain this briefly.